optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py

```diff
@@ -334,18 +335,39 @@ def forward(
 
         key, value = key_value.split((self.head_dim, self.head_dim), dim=-1)
 
-        if layer_past is not None:
+        _, q_len, _ = hidden_states.size()
+        bucket_internal_decode_stage = cache_idx is not None and q_len == 1
+
+        if not bucket_internal_decode_stage:
+            if layer_past is not None:
+                past_key, past_value = layer_past.split((self.head_dim, self.head_dim), dim=-1)
+                if token_idx is not None:
+                    # Using out of place version of index_add_() to ensure the intermediate tensors are not lost when HPU graphs are enabled.
+                    key = past_key.index_add(1, token_idx - 1, key - torch.index_select(past_key, 1, token_idx - 1))
+                    value = past_value.index_add(
+                        1, token_idx - 1, value - torch.index_select(past_value, 1, token_idx - 1)
+                    )
+                else:
+                    key = torch.cat((past_key, key), dim=-2)
+                    value = torch.cat((past_value, value), dim=-2)
+            present = torch.cat((key, value), dim=-1) if use_cache else None
+        else:
+            assert (
+                token_idx is not None
+            ), "Invalid parameters: token_idx is None at decode stage with bucket_internal"
+            assert (
+                layer_past is not None
+            ), "Invalid parameters: layer_past is None at decode stage with bucket_internal"
+
             past_key, past_value = layer_past.split((self.head_dim, self.head_dim), dim=-1)
-            if token_idx is not None:
-                # Using out of place version of index_add_() to ensure the intermediate tensors are not lost when HPU graphs are enabled.
-                key = past_key.index_add(1, token_idx - 1, key - torch.index_select(past_key, 1, token_idx - 1))
-                value = past_value.index_add(
-                    1, token_idx - 1, value - torch.index_select(past_value, 1, token_idx - 1)
-                )
-            else:
-                key = torch.cat((past_key, key), dim=-2)
-                value = torch.cat((past_value, value), dim=-2)
-        present = torch.cat((key, value), dim=-1) if use_cache else None
+            key = past_key.index_copy_(1, token_idx - 1, key)
```

Comment:
can you verify this works with tgi-gaudi.. out of place op was used to fix a specific issue when tensor cache is disabled otherwise we saw error

Reply:
sent you ticket link of empty tensor optional error with tgi-gaudi

Reply:
@vidyasiv I tried to rollback changes from your commit https://github.com/huggingface/optimum-habana/pull/1181 and it works for me on latest 1.18 with command line ```PT_HPU_DISABLE_TENSOR_CACHE=1 python run_generation.py --model_name_or_path bigcode/starcoder --batch_size 2 --use_hpu_graphs --use_kv_cache  --max_new_tokens 100 --bf16``` and output is the same as without ```PT_HPU_DISABLE_TENSOR_CACHE``` variable. It seems that original issues was fixed

Reply:
for bucket it also works fine and gives the same output:
```
PT_HPU_DISABLE_TENSOR_CACHE=1 python run_generation.py --model_name_or_path bigcode/starcoder --batch_size 2 --use_hpu_graphs --use_kv_cache  --max_new_tokens 100 --bf16 --bucket_size=128 --bucket_internal
```

Reply:
@vidyasiv What's the TGI config that was leading to an error?